File: dotnet10-dapper/Dotnet10Dapper/Repository/FruitRepository.cs
using System.Data;
using Dapper;
using Dotnet10Dapper.Dto;
using Npgsql;

namespace Dotnet10Dapper.Repository;

public sealed class FruitRepository(NpgsqlDataSource dataSource) : IFruitRepository
{
    private const string SelectSql = """
        SELECT f.id AS FruitId, f.name AS FruitName, f.description AS Description,
               p.price AS Price, s.id AS StoreId, s.name AS StoreName,
               s.currency AS Currency, s.address AS Address, s.city AS City,
               s.country AS Country
          FROM fruits f
          LEFT JOIN store_fruit_prices p ON p.fruit_id = f.id
          LEFT JOIN stores s ON s.id = p.store_id
        """;

    public async Task<List<FruitDto>> ListAllAsync(CancellationToken cancellationToken = default)
    {
        await using var connection = await dataSource.OpenConnectionAsync(cancellationToken);
        var rows = await connection.QueryAsync<FruitRow>(
            new CommandDefinition(SelectSql + " ORDER BY f.id, s.id", cancellationToken: cancellationToken));
        return Map(rows);
    }

    public async Task<FruitDto?> FindByNameAsync(string name, CancellationToken cancellationToken = default)
    {
        await using var connection = await dataSource.OpenConnectionAsync(cancellationToken);
        var rows = await connection.QueryAsync<FruitRow>(new CommandDefinition(
            SelectSql + " WHERE f.name = @Name ORDER BY s.id", new { Name = name },
            cancellationToken: cancellationToken));
        return Map(rows).FirstOrDefault();
    }

    public async Task<FruitDto> PersistAsync(FruitDto fruit, CancellationToken cancellationToken = default)
    {
        await using var connection = await dataSource.OpenConnectionAsync(cancellationToken);
        await using var transaction = await connection.BeginTransactionAsync(cancellationToken);
        try
        {
            var id = await connection.ExecuteScalarAsync<long>(new CommandDefinition(
                """INSERT INTO fruits (id, name, description) VALUES (nextval('fruits_seq'), @Name, @Description) RETURNING id""",
                new { fruit.Name, fruit.Description }, transaction, cancellationToken: cancellationToken));
            await transaction.CommitAsync(cancellationToken);
            return new FruitDto(id, fruit.Name, fruit.Description, []);
        }
        catch
        {
            await transaction.RollbackAsync(cancellationToken);
            throw;
        }
    }

    private static List<FruitDto> Map(IEnumerable<FruitRow> rows)
    {
        var fruits = new Dictionary<long, FruitAccumulator>();
        foreach (var row in rows)
        {
            if (!fruits.TryGetValue(row.FruitId, out var fruit))
            {
                fruit = new FruitAccumulator(row.FruitId, row.FruitName, row.Description);
                fruits.Add(row.FruitId, fruit);
            }
            if (row.StoreId is not null)
            {
                var address = new AddressDto(row.Address!, row.City!, row.Country!);
                var store = new StoreDto(row.StoreId, row.StoreName!, row.Currency!, address);
                fruit.StorePrices.Add(new StoreFruitPriceDto(store, (float)row.Price!.Value));
            }
        }
        return fruits.Values.Select(f => new FruitDto(f.Id, f.Name, f.Description, f.StorePrices)).ToList();
    }

    private sealed record FruitRow(long FruitId, string FruitName, string? Description,
        decimal? Price, long? StoreId, string? StoreName, string? Currency,
        string? Address, string? City, string? Country);

    private sealed record FruitAccumulator(long Id, string Name, string? Description)
    {
        public List<StoreFruitPriceDto> StorePrices { get; } = [];
    }
}

File: helidon4-mp-jooq/src/main/java/org/acme/helidon/mp/repository/FruitRepository.java
package org.acme.helidon.mp.jooq.repository;

import static org.jooq.impl.DSL.field;
import static org.jooq.impl.DSL.name;
import static org.jooq.impl.DSL.table;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Optional;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import org.acme.helidon.mp.jooq.dto.AddressDTO;
import org.acme.helidon.mp.jooq.dto.FruitDTO;
import org.acme.helidon.mp.jooq.dto.StoreDTO;
import org.acme.helidon.mp.jooq.dto.StoreFruitPriceDTO;
import org.jooq.DSLContext;
import org.jooq.Field;
import org.jooq.Record;
import org.jooq.Result;
import org.jooq.Table;

@ApplicationScoped
public class FruitRepository {
    private static final Table<?> FRUITS = table(name("fruits")).as("f");
    private static final Table<?> PRICES = table(name("store_fruit_prices")).as("p");
    private static final Table<?> STORES = table(name("stores")).as("s");
    private static final Field<Long> FRUIT_ID = field(name("f", "id"), Long.class);
    private static final Field<String> FRUIT_NAME = field(name("f", "name"), String.class);
    private static final Field<String> DESCRIPTION = field(name("f", "description"), String.class);
    private static final Field<Long> PRICE_FRUIT_ID = field(name("p", "fruit_id"), Long.class);
    private static final Field<Long> PRICE_STORE_ID = field(name("p", "store_id"), Long.class);
    private static final Field<BigDecimal> PRICE = field(name("p", "price"), BigDecimal.class);
    private static final Field<Long> STORE_ID = field(name("s", "id"), Long.class);
    private static final Field<String> STORE_NAME = field(name("s", "name"), String.class);
    private static final Field<String> CURRENCY = field(name("s", "currency"), String.class);
    private static final Field<String> ADDRESS = field(name("s", "address"), String.class);
    private static final Field<String> CITY = field(name("s", "city"), String.class);
    private static final Field<String> COUNTRY = field(name("s", "country"), String.class);

    private final DSLContext dsl;

    @Inject
    public FruitRepository(DSLContext dsl) {
        this.dsl = dsl;
    }

    public List<FruitDTO> findAll() {
        return map(selectFruits().orderBy(FRUIT_ID, STORE_ID).fetch());
    }

    public Optional<FruitDTO> findByName(String fruitName) {
        return map(selectFruits().where(FRUIT_NAME.eq(fruitName)).orderBy(STORE_ID).fetch())
                .stream().findFirst();
    }

    public FruitDTO insert(String fruitName, String description) {
        return dsl.transactionResult(configuration -> {
            DSLContext tx = configuration.dsl();
            Long id = tx.select(field("nextval('fruits_seq')", Long.class)).fetchOne(0, Long.class);
            tx.insertInto(table(name("fruits")), field(name("id")), field(name("name")),
                            field(name("description")))
                    .values(id, fruitName, description)
                    .execute();
            return new FruitDTO(id, fruitName, description, List.of());
        });
    }

    private org.jooq.SelectOnConditionStep<? extends Record> selectFruits() {
        return dsl.select(FRUIT_ID, FRUIT_NAME, DESCRIPTION, PRICE, STORE_ID, STORE_NAME,
                        CURRENCY, ADDRESS, CITY, COUNTRY)
                .from(FRUITS)
                .leftJoin(PRICES).on(PRICE_FRUIT_ID.eq(FRUIT_ID))
                .leftJoin(STORES).on(STORE_ID.eq(PRICE_STORE_ID));
    }

    static List<FruitDTO> map(Result<? extends Record> rows) {
        LinkedHashMap<Long, FruitAccumulator> fruits = new LinkedHashMap<>();
        for (Record row : rows) {
            Long fruitId = row.get(FRUIT_ID);
            FruitAccumulator fruit = fruits.computeIfAbsent(fruitId,
                    ignored -> new FruitAccumulator(fruitId, row.get(FRUIT_NAME), row.get(DESCRIPTION)));
            Long storeId = row.get(STORE_ID);
            if (storeId != null) {
                StoreDTO store = new StoreDTO(storeId, row.get(STORE_NAME), row.get(CURRENCY),
                        new AddressDTO(row.get(ADDRESS), row.get(CITY), row.get(COUNTRY)));
                fruit.prices.add(new StoreFruitPriceDTO(store, row.get(PRICE).floatValue()));
            }
        }
        return fruits.values().stream().map(FruitAccumulator::toDto).toList();
    }

    private static final class FruitAccumulator {
        private final Long id;
        private final String name;
        private final String description;
        private final List<StoreFruitPriceDTO> prices = new ArrayList<>();

        private FruitAccumulator(Long id, String name, String description) {
            this.id = id;
            this.name = name;
            this.description = description;
        }

        private FruitDTO toDto() {
            return new FruitDTO(id, name, description, prices);
        }
    }
}
